manubot/pandoc/cite_filter.py

```diff
@@ -0,0 +1,219 @@
+    manuscript_citekeys.append(elem.id)
+    return None
+
+
+def _citation_to_id_action(elem, doc):
+    """
+    Panflute action to update the citationId of Citations in the AST
+    with their manubot-created keys.
+    """
+    if not isinstance(elem, pf.Citation):
+        return None
+    mapper = global_variables["citekey_shortener"]
+    if elem.id in mapper:
+        elem.id = mapper[elem.id]
+    return None
+
+
+def _get_reference_link_citekey_aliases(elem, doc):
+    """
+    Based on TypeScript implementation by phiresky at
+    https://github.com/phiresky/pandoc-url2cite/blob/b28374a9a037a5ce1747b8567160d8dffd64177e/index.ts#L118-L152
+
+    Uses markdown's link reference syntax to define citekey aliases (tags)
+    https://spec.commonmark.org/0.29/#link-reference-definitions
+    """
+    if type(elem) != pf.Para:
+        return
+    while (
+        len(elem.content) >= 3
+        and type(elem.content[0]) == pf.Cite
+        and len(elem.content[0].citations) == 1  # differs from pandoc-url2cite
+        and type(elem.content[1]) == pf.Str
+        and elem.content[1].text == ":"
+    ):
+        space_index = 3 if type(elem.content[2]) == pf.Space else 2
+        destination = elem.content[space_index]
+        if type(destination) == pf.Str:
+            # paragraph starts with [@something]: something
+            # save info to citekeys and remove from paragraph
+            citekey = elem.content[0].citations[0].id  # differs from pandoc-url2cite
+            citekey_aliases = global_variables["citekey_aliases"]
+            if citekey in citekey_aliases:
+                logging.warning(f"duplicate citekey {citekey}")
+            citekey_aliases[citekey] = destination.text
+            # found citation, add it to citekeys and remove it from document
+            elem.content = elem.content[space_index + 1 :]
+        if len(elem.content) > 0 and type(elem.content[0]) == pf.SoftBreak:
+            elem.content.pop(0)
+
+
+def process_citations(doc):
+    """
+    Apply citation-by-identifier to a Python object representation of
+    Pandoc's Abstract Syntax Tree.
+
+    The following Pandoc metadata fields are considered:
+
+    - bibliography (use to define reference metadata manually)
+    - citekey-aliases (use to define tags for cite-by-id citations)
+    - manubot.requests-cache-path
+    - manubot.clear-requests-cache
+    - manubot.output-bibliography: path to write generated CSL JSON bibliography
+    """
+    citekey_aliases = doc.get_metadata("citekey-aliases", default={})
+    if not isinstance(citekey_aliases, dict):
+        logging.warning(
+            f"Expected metadata.citekey-aliases to be a dict, "
+            f"but received a {citekey_aliases.__class__.__name__}. Disregarding."
+        )
+        citekey_aliases = dict()
+
+    global_variables["citekey_aliases"] = citekey_aliases
+
+    manuscript_citekeys = set(global_variables["manuscript_citekeys"])
```

Comment:
Why does this need to be called before the two `walk` calls?

Reply:
good call. this line is completely inert as far as i can tell